### ISSUE TITLE
Timebox: Proof of concept for tracking references in lambdas

### ISSUE THREAD

**Issue by jsternberg** [influxdata/flux#3877]
This issue was originally part of #3865 and is being split into a new issue.

We would like to improve the efficiency of the Flux engine by reducing the amount of memory needed and number of allocations performed. One way to do this is to perform reference counting on tables and or columns (arrow arrays) within those tables.

Consider for example that the output of many transformations might leave an input column untouched. For example, in this very common pattern:

...
  |> map(fn: (r) => ({r with newCol: r._value * 10}))
...
All of the columns in the input tables to map will be output unchanged. If we could reuse the input columns in the output, we would save the overhead of additional allocations and needless copying.

**Comment by jsternberg** [comment]
Comment copied:

> Consider for example that the output of many transformations might leave an input column untouched. For example, in this very common pattern:
>
>     ...
>       |> map(fn: (r) => ({r with newCol: r._value * 10}))
>     ...

This one is interesting because I think it touches a similar place as vectorization. In this case, the map function has some semantic graph and it needs to determine the meaning of the semantic graph. It needs to determine which values are accessed, transformed, and ultimately output. At the current moment, our interpreter runtime (in code, it's called the compiler and is in the compiler package) is very naive about this and ultimately uses a lot of memory to capture what is happening. The `{}` creates a new object and allocates memory each time that the function is executed. The `map()` function suffers from this more because object creation is a lot of small allocations. In general, the functions passed to filter end up being much faster because they don't allocate as much memory.

There's a few major components to this:

1. Code evaluation should be more efficient. We have enough information that we can properly compile this statement rather than interpreting it from semantic graph.
2. Memory allocation does not need to be done for each row. We know what `r` contains for the processing of a table with a specific group key and we don't need to repeat this allocation for every row. We can reuse the same memory within the runtime stack. I'm not referring to arrow arrays here. I'm referring to the equivalent of registers/stack memory.
3. I think we'll need to probably design an IR language to capture program flow. If we rely on processing semantic graph, we're going to miss a lot of optimization opportunities. This is the kind of analysis that SSA IR was designed for. For example, the following is equivalent to the above:

```
myfunc = (r) => {
    val = r._value * 10
    return {r with newCol: val}
}
```

But if we relied solely on the present semantic graph, I think we'd start to have difficulty. Instead, I think we need to consider transforming this into an SSA IR. The above two would convert into the same IR.

```
define record @myfunc(record %r) {
    %1 = getvalue %r, "_value"
    %2 = loadconst 10
    %3 = mul %1, %2
    %4 = extend %r, "newCol", %3
    ret %4
}
```

While SSA IR's are traditionally output as assembly, they naturally convert directly into the equivalent of our planning graph. I suspect this will open up more opportunities for modifying the IR for optimization. It also has a few other side benefits:

1. It's no longer possible for a person to break the stack by having 1,000 conditions since processing the function will no longer involve any recursion.
2. Recursion isn't needed for analyzing or modifying the function.

This is very abbreviated because I'm just trying to get some thoughts down. I'll try to expand more. I might have to do some digging in to see if my thoughts lead anywhere useful.

**Comment by jsternberg** [comment]
Comment from @wolffcm:

I've been thinking about this. I think it would be possible to be able to do the analysis we need  (as to which values are accessed, transformed etc), leverage vectorization, and still use our existing semantic graph as the representation. 

Consider a call to map with a lambda, where the fields of `r` are all floats:
```
map(fn: (r) => ({_value: r.a * r.b + r.c}))
```
With vectorization, we want to be able to deal with fields of `r` as vectors, rather than individual values. We could transform this lambda with a `Vectorization` code transformation like this:
```
(_r_a, _r_b, _r_c) => ({_value: _r_a * _r_b + _r_c})
```
Where the types of `_r_a` etc are `[float]`, and the return type is `{_value: [float]}`. 

The steps for this code transformation would be something like this:
1. We start knowing the actual concrete type of `r` and all its fields
2. Replace any instances of `{r with ...}` with a record expression that explicitly includes all fields
3. Replace the `r` parameter with the individual parameters for each field that is referenced via a `MemberExpression`. 
4. Replace each `MemberExpression` that uses `r` with an `IdentifierExpression` like `_r_a`, etc.

Then we could evaluate the lambda by calling it with arrays exposed by the `ColReader`.

By exploding the fields of the record out into individual arrays, we get closer to the vector operations we actually want to perform, and it also becomes easier to analyze how each array input is used (we might consider implementing a rudimentary version of [use def chains](https://en.wikipedia.org/wiki/Use-define_chain)). Note that semantic graph is already in SSA form; variables can only be assigned once statically.

We would need to make the compiler able to support elementwise vector operations. And we may wish to extend the semantic graph representation itself to make elementwise operations explicit, by adding new expression types like `BinaryElementwiseExpression`, `CallElementwiseExpression` and so on.

I don't see a whole lot of value in a three-address-code style IR except where we are concerned about stack depth and recursion. But since Go can grow the stack dynamically, and we have alternatives like iterative traversal using stacks or queues allocated on the heap, it seems a minor concern, unless I'm missing something.

**Comment by jsternberg** [comment]
I think I may understand more what you mean by vectorization. You want to transform the lambda expression to be from each individual row to something that operates on columns and performs operations on the columns themselves rather than on individual values.

> I've been thinking about this. I think it would be possible to be able to do the analysis we need (as to which values are accessed, transformed etc), leverage vectorization, and still use our existing semantic graph as the representation.

Honestly, our semantic graph is very close to an IR but I think it misses a few key components that prevent it from being a suitable IR for analysis and optimization. I'd like to explore these as if we can reuse the semantic graph as the IR then I'm all for doing that.

Some key components that I think end up being problematic with semantic graph:

1. Very difficult to represent in a textual format. This is most important for developers to be able to understand a program flow, debug, and then iterate on optimizations. This is typically represented like the LLVM-like IR that I put into one of the above comments. While this may seem extra, I think that we need a way to represent the semantic meaning of a section of code in a textual format that is easy for a human to read. Otherwise, I think we'll have too much difficulty in communicating with each other and understanding ourselves.
2. The "graph" is not a graph but a recursive semi-normalized AST. IR transformations for optimization or normalization are usually designed around a graph.
3. It's not actually normalized and not actually SSA. While there is single assignment because flux is single assignment, the semantic graph still references identifiers and indirect accesses. There's also multiple ways to represent the same thing. For example, see the following:
    ```
    value_gt = (r) => r._value > 0
    ...
    |> filter(fn: (r) => r._value > 0)
    |> filter(fn: value_gt)
    ```
    These filters are identical, but the current semantic graph will represent one as an identifier expression and one as a function expression. A properly normalized version, such as a proper IR, would instantiate the anonymous function and these two would look much more similar in how they are invoked.

    ```
    define bool @value_gt(record %r) {
        %1 = getvalue %r, "_value"
        %2 = loadconst 0
        %3 = cmp_gt %1, %2
        ret %3
    }

    define bool @main.1(record %r) {
        %1 = getvalue %r, "_value"
        %2 = loadconst 0
        %3 = cmp_gt %1, %2
        ret %3
    }

    define void main() {
        %1 = call_builtin "from", ...
        %2 = call_builtin "filter", %1, @value_gt
        %3 = call_builtin "filter", %1, @main.1
    }
    ```

    For the above, it's trivial to determine that these two functions are identical and it is also much more trivial to run the same optimization on both.

    I suspect it will be harder to normalize the current semantic graph than we anticipate. I think there will be a lot of cases we haven't thought about and I think the human-readable output format will be needed for any endeavor where we try.

> I don't see a whole lot of value in a three-address-code style IR except where we are concerned about stack depth and recursion. But since Go can grow the stack dynamically, and we have alternatives like iterative traversal using stacks or queues allocated on the heap, it seems a minor concern, unless I'm missing something.

For what it's worth, I don't think strict three-address-code style is needed. For the IR, I think it's mostly helpful for human consumption. The real value of an IR is:

1. It provides a human-readable way to debug and observe the system.
2. It provides a natural method to perform normalizations on the AST.
3. It translates 1:1 into a directed graph structure.

The directed graph structure is really the only one that affects the actual computation as the directed graph structure is where pattern matching and normalizations occur. Optimizations are done by either matching certain patterns in the graph structure or traversing the graph and performing specific transformations. A recursive structure can make this difficult to do.

Similarly, I wouldn't completely dismiss the effect of recursion on both performance and stability. While Go can dynamically grow the stack, this isn't infinite and it isn't without a performance penalty. It still has to allocate memory for each level of the stack and resizing the stack isn't without penalties itself. Similarly, the Rust section of the code cannot dynamically reallocate the stack which limits the maximum number of conditions.

**Comment by wolffcm** [comment]
> It's not actually normalized and not actually SSA. While there is single assignment because flux is single assignment, the semantic graph still references identifiers and indirect accesses. 

This doesn't mesh with my understanding of what's required for SSA form. Our semantic graph might not look like an LLVM IR, but since each of our identifier expressions are defined in exactly one place in the program it's SSA. Could you clarify what you mean by "indirect accesses" though?

I'm also curious about what the benefits of normalization would be. Are you talking about optimizations like common subexpression elimination and the like?

> [Semantic graph is] Very difficult to represent in a textual format

Is there any reason we couldn't write a formatter for semantic graph to create a textual format, and perhaps a parser to go the other way? The main difference between AST and semantic graph is the type annotations on expressions. We could put them into the textual representation.

> The "graph" is not a graph but a recursive semi-normalized AST

It's not really a graph, but there are edges implied by the lists of statements in the top level or in function bodies. We don't really have any control flow statements, so modeling the edges explicitly seems redundant. Even our conditional expressions are at the expression level. (But there is the significant detail that conditional expressions do short circuit so side effects in the branch not taken do not occur.)

> Similarly, I wouldn't completely dismiss the effect of recursion on both performance and stability

I don't mean to dismiss them. It would be good to get rid of any possibility of unbounded recursion. Even a graph-based IR would lend itself to recursive traversal, and we would need to assiduously avoid it there too. I think we should work to remove unbounded recursion, but I just don't see it as a priority at the moment.

I guess I would dispute that semantic graph is not a "proper" IR. It could be improved, but it seems like an IR to me, and most of the optimizations that we're currently considering (vectorization, tracking array references) seem possible with it.

With that said, longer term we might want such a representation, especially if we plan on emitting native code from Flux. It would certainly be a fun project to work on.

**Comment by jsternberg** [comment]
I'm going to assign myself to experiment on this, but anybody else who wants to work on this feel free to pick it up. I think there are going to be multiple ways to approach this.

@wolffcm going to write a different comment to continue the conversation.

**Comment by jsternberg** [comment]
> This doesn't mesh with my understanding of what's required for SSA form. Our semantic graph might not look like an LLVM IR, but since each of our identifier expressions are defined in exactly one place in the program it's SSA. Could you clarify what you mean by "indirect accesses" though?

I think what I mean by this is that there are multiple ways to access the same value. For example, say that we want to identify a specific call expression like `strings.strlen()` inside of a function. We can access this through:

```
import "strings"
length = strings.strlen
|> map(fn: (r) => ({r with len: strings.strlen(v: r._measurement)}))
|> map(fn: (r) => ({r with len: length(v: r._measurement)}))
```

These are the same, but with our current compiler runtime, we wouldn't be able to identify themselves as the same statically. If I wanted to write Go code to match a function expression, I would have to do a switch statement on the expression and exhaust the possible different expression nodes.

For example, here's the `CallExpression` struct (translating from flatbuffers to what it would look like in Go and removing certain attributes that aren't relevant).

```go
type CallExpression struct {
    Callee Expression
    Arguments []Property
    Pipe Expression
}
```

If I wanted to write an expression to match on the above, I'd have to evaluate the callee. If I wanted to match on arguments or the pipe expression, I'd have to expand those two. Now we could do steps to normalize the call expression. For example, we can rewrite all call expressions to always have an identifier in `Callee` and move the pipe argument to the appropriate property in arguments, but I think there's a lot of examples where this could be difficult.

I just think this would end up being too difficult over time to maintain and I think that an IR doubles as a useful way to think about these problems. My view of an IR is that each IR instruction can only have one possible input type so it would be easier to perform the matches.

Here's a weird example that I think might end up being difficult to evaluate with the current semantic graph.

```
compare = (r, fn) => fn(v: r._value)

...
|> filter(fn: (r) => compare(r: r, fn: (v) => v > 0))
```

Attempting to represent this in semantic graph, I think this would be it (only doing the filter function expression):

```
FunctionExpression{
    Fn: CallExpression{
        Callee: Identifier{"compare"},
        Arguments: [
            R: Identifier{"r"},
            Fn: FunctionExpression{
                Fn: BinaryExpression{
                    L: Identifier{"v"},
                    R: Literal{0}
                }
            }
        ]
    }
}
```

This doesn't include compare itself. We reference compare as the callee in the function expression but I had a difficult time putting it in here.

In comparison, I think if we redid the semantic graph with the idea of an IR, we'd end up with something more readable that would be easier to modify. I think the primary benefit is it's non-recursive and I think non-recursive structures with fewer options are easier to reason with.

```
define bool @compare(record %r, func %fn) {
    %1 = call_indirect %fn, %r
    ret %1
}

define bool @filter.1(int %v) {
    %1 = loadconst 0
    %2 = cmp_gt %v, %1
    ret %2
}

define bool @main.1(record %r) {
    %1 = call @compare, %r, @filter.1
    ret %1
}
```

I find it easy to look at the above and see that we can optimize `call_indirect` to be a direct call because `@main.1` invokes `@compare` with a static function. I think these same optimizations would work with the semantic graph, but I think it would be harder to both see, reason, and modify the program without adding bugs and I think these are requirements when attempting these optimizations.

> Is there any reason we couldn't write a formatter for semantic graph to create a textual format, and perhaps a parser to go the other way? The main difference between AST and semantic graph is the type annotations on expressions. We could put them into the textual representation.

In my opinion, recursive structures are very difficult to follow. I tried to translate the above program into a recursive structure and I'm pretty sure I messed up. I was working on writing it for some time before I just gave up and said "good enough I don't have to validate this". If I were to show you a recursive call chain and convert it into a graph, both equivalent, which of these would you have an easier time reasoning with?

```
E(D(C(B(A))))
A -> B -> C -> D -> E
```

I actually had a difficult time writing the first one because I had a difficult time counting parenthesis and this was a simple pipeline. If I did this:

```
E(D(C),B(A))
A -> B
        \
C -> D -> E
```

Both of these are equivalent, but the graph is easier to read. I have to try it myself, but I think it's also easier to convert into a procedural ASM.

**Comment by jsternberg** [comment]
Some notes on what I've experimented with so far.

I managed to convert the semantic graph into a kind of simplified IR graph. I then formatted the output into a digraph and also created a formatter to output it as an ASM-style IR.

Some things that I found:

* My first attempt was to parse into an ASM-style IR. There didn't seem to be much value to this.
* The graph style IR felt like it was useful. In particular, I was able to create the graph rather easily and then reuse the planner code to match certain patterns.
* The graph style seems to have simplified some of the semantic graph for output. In particular, the following two queries are different semantic graph and end up with the same IR graph.

    ```
    fn = (r) => {
        a = {r with a: 1}
        return a
    }

    fn = (r) => ({r with a: 1})
    ```
* It's unclear just how much of a difference there is between semantic graph and this IR graph. I'm concerned that it would still be easy to use the current semantic graph instead of creating yet another transformation, but the graph format does seem to create some clear advantages. On the other hand, I don't really understand what benefits the current semantic graph gives.

A clear benefit of switching the data structure was that I was able to reuse the planner code to modify the graph and extraneous information such as parenthesis or identifiers get removed entirely. I suspect it will also be easier to modify the actual graph too but I don't really have any evidence of that.

The concern I have is whether this additional transformation provides enough benefits to be worth implementing. I'll be trying to use it to optimize map and see what happens.

My next steps:

* Include type information in the IR graph.
* Modify the map transformation to construct the IR graph and trace which columns are passthroughs.

**Comment by jsternberg** [comment]
I think I've gotten most of what I'm going to get out of this timebox so I'm going to try and record some notes. I don't have very clear action items but I do have some general directions if we decide to go this kind of way into optimizing map calls.

I suspect that the optimizations that can be made here from preserving columns through a map call are probably less than other optimizations that can be made to map so it is my opinion we should revisit map performance in other locations before we look into tracking lambda references. In particular, the compiler runtime could use optimization itself and map is still using the older column list table builder and would likely benefit from changing from that.

I identified the following as requirements to be able to track lambdas going through the map function to identify which ones would not be modified.
First, we need to be able to identify the return node as an object expression or a reference to the input parameter.
Second, we need to determine the inputs to that object expression.
We need to correspond the type of the object expression as determined by type inference to the input parameters.
This may involves substituting types because type inference will not have given us all of the required information.
Third, we need to identify if the object expression extends a record and that record is a descendent of the input parameter.
Fourth, we need to identify any input parameters to the object expression of the form `{a: r["xxx"]}`.

This information is all available in the semantic graph and most of the code involved with doing this is in various places, but they are all a bit scattered between different portions of the code.
For example, we have the code for substituting types, but we don't have code for substituting types and then reassigning them to the semantic graph node that type is associated with.
We have the ability to determine if the referenced record is the input parameter, but we don't have a way to track it through reassignments to different identifiers because the semantic graph is not normalized for variable references and assignments.

We also have some small problems in the existing code that would require changes.
Substitute types seems to mostly work, but it seems to be designed primarily around the current compiler runtime.
For records, it will prune and sort duplicates when it is substituting types.
For this particular transformation, it needs to return exactly what type inference would have returned.
The reason for this is because we need to track the `{r with ...}` keyword and identify which of the columns from `r` are getting shadowed.
Without knowing which `_value` is being referenced after type substitution, we can't know that the `_value` from `r` is being shadowed by a new column.

I identified the following problems with our current semantic graph that make some of the above difficult.

The most significant is probably that the semantic graph is not normalized.
There are multiple ways to write the same semantic meaning to a program.
For example, the following result in the same program and different semantic graphs:

    fn = (r) => ({r with a: 1})

    fn = (r) => {
      a = 1
      return {r with a: a}
    }

It is pretty easy to unwind this, but is a generally required step to ensure that the pattern is correctly identified.
After we normalize it, we need to keep a mapping of identifiers to their instruction which requires additional storage and steps to analyze a function block.
If we choose to normalize it by removing identifiers instead, we cannot represent programs that have one node with multiple successors such as the following:

    fn = (r) => {
      a = r._value
      return {r with a: a, b: a}
    }

This example is a bit contrived, but it is valid flux and needs to be represented in a normalized graph.

If we were to duplicate the branch to avoid multiple successors, we could potentially duplicate function calls or other more complex code that shouldn't be duplicated.

A secondary problem is that transformations on the semantic graph are harder because it can be more difficult to update references.
Here is a normalized lambda that we can optimize:

    fn = (r) => {
      b = {r with a: 1}
      return {b with b: 2}
    }

We can optimize these by combining the object expressions into a single one.

    fn = (r) => {
      return {r with a: 1, b: 2}
    }

But if `b` is referenced from another node in this same block, we can potentially change the behavior of the code.
It's necessary for us to understand how the data passes to successors of a particular node before making some optimizations.

A tertiary problem is that the semantic graph is one directional because it uses recursive structures.
This means that any identifying patterns can only go backwards.
This generally isn't a large problem, but can become a problem with certain optimizations such as the one above.
We can only make that optimization if we can identify that `b` is not being used by another node.
If `b` is being used by another node, we can't collapse the calls.
